File: frontend/src/data/grammar.ts
export type { CEFRLevel, GrammarCategory } from '@/data/types'

import { apiFetch } from '@/lib/api'

export interface GrammarExample {
  text: string
  translation?: string
  note?: string
}

export interface GrammarMistake {
  wrong: string
  correct: string
  note: string
}

export interface GrammarTopic {
  slug: string
  title: string
  level: string
  category: string
  summary: string
  explanation: string
  structure?: string
  rules: string[]
  examples: GrammarExample[]
  common_mistakes: GrammarMistake[]
  related: string[]
}

export interface GrammarNativeHelpExample {
  sentence: string
  note: string
}

export interface GrammarNativeHelpTrap {
  mistake: string
  fix: string
}

export interface GrammarNativeHelpGlossaryItem {
  term: string
  meaning: string
  note?: string
}

export interface GrammarNativeHelp {
  summary: string
  explanation: string
  key_points: string[]
  examples: GrammarNativeHelpExample[]
  common_traps: GrammarNativeHelpTrap[]
  mini_glossary: GrammarNativeHelpGlossaryItem[]
}

export async function getGrammarTopics(
  targetLanguage: string = 'en-GB'
): Promise<GrammarTopic[]> {
  const res = await apiFetch(
    `/api/grammar?language=${encodeURIComponent(targetLanguage)}`
  )
  if (!res.ok) return []
  const data = await res.json()
  return data.topics ?? []
}

export async function getGrammarNativeHelp(
  slug: string,
  targetLanguage: string = 'en-GB'
): Promise<GrammarNativeHelp | null> {
  const res = await apiFetch(
    `/api/grammar/${encodeURIComponent(slug)}/native-help?language=${encodeURIComponent(targetLanguage)}`,
    { method: 'POST' }
  )
  if (!res.ok) return null
  const data = await res.json()
  return data.native_help ?? null
}

File: backend/app/schemas/grammar.py
from __future__ import annotations

from pydantic import BaseModel


class GrammarExampleResponse(BaseModel):
    text: str
    translation: str | None = None
    note: str | None = None


class GrammarMistakeResponse(BaseModel):
    wrong: str
    correct: str
    note: str


class GrammarTopicResponse(BaseModel):
    slug: str
    title: str
    level: str
    category: str
    summary: str
    explanation: str
    structure: str | None = None
    rules: list[str]
    examples: list[GrammarExampleResponse]
    common_mistakes: list[GrammarMistakeResponse]
    related: list[str]


class GrammarTopicsResponse(BaseModel):
    topics: list[GrammarTopicResponse]


class GrammarTopicDetailResponse(BaseModel):
    topic: GrammarTopicResponse


class GrammarNativeHelpExampleResponse(BaseModel):
    sentence: str
    note: str


class GrammarNativeHelpTrapResponse(BaseModel):
    mistake: str
    fix: str


class GrammarNativeHelpGlossaryResponse(BaseModel):
    term: str
    meaning: str
    note: str | None = None


class GrammarNativeHelpContentResponse(BaseModel):
    summary: str
    explanation: str
    key_points: list[str]
    examples: list[GrammarNativeHelpExampleResponse]
    common_traps: list[GrammarNativeHelpTrapResponse]
    mini_glossary: list[GrammarNativeHelpGlossaryResponse]


class GrammarNativeHelpResponse(BaseModel):
    native_help: GrammarNativeHelpContentResponse
